src/esphome/sensor/mhz19_component.cpp

```diff
@@ -26,18 +27,59 @@ uint8_t mhz19_checksum(const uint8_t *command) {
 
 void MHZ19Component::update() {
   uint8_t response[MHZ19_RESPONSE_LENGTH];
+repeat:
   if (!this->mhz19_write_command_(MHZ19_COMMAND_GET_PPM, response)) {
     ESP_LOGW(TAG, "Reading data from MHZ19 failed!");
     this->status_set_warning();
     return;
   }
 
   if (response[0] != 0xFF || response[1] != 0x86) {
-    ESP_LOGW(TAG, "Invalid preamble from MHZ19!");
+    ESP_LOGW(TAG, "Invalid preamble from MHZ19! [%0x %0x %0x %0x %0x %0x %0x %0x %0x]",
+      response[0], response[1], response[2], response[3], response[4],
+      response[5], response[6], response[7], response[8]);
     this->status_set_warning();
+    if (++this->cleanup_rx_buffer > 2) {
+        /* try to recover and give up on it if several attemprs were no successful until the next update cycle */
+        this->cleanup_rx_buffer = 0;
+        return;
+    }
+    goto repeat;
+  }
+  this->cleanup_rx_buffer = 0;
+
+  /* Sensor reports U(15000) during boot, ingnore reported CO2 until it boots */
+  uint16_t u = (response[6] << 8) + response[7];
+  if (u == 15000) {
+    ESP_LOGD(TAG, "Sensor is booting");
     return;
   }
 
+  /* MH-Z19B(s == 0) and MH-Z19(s != 0) */
+  uint8_t s = response[5];
+  if (response[5] == 0 && this->model_b == false) {
+    ESP_LOGD(TAG, "MH-Z19B detected");
+    this->model_b = true;
+  }
+
+  if (this->model_b && this->abc_disabled == false) {
+    uint8_t abc_ack[MHZ19_RESPONSE_LENGTH];
+    /*
+     * MH-Z19B allows to enable/disable 'automatic baseline calibration' (datasheet MH-Z19B v1.2),
+     * disable it to prevent sensor baseline drift in not well ventilated area
+     */
+    ESP_LOGI(TAG, "Disabling ABC on boot");
+    if (!this->mhz19_write_command_(MHZ19_COMMAND_ABC_DISABLE, abc_ack)) {
+      ESP_LOGW(TAG, "Failed to read ABC disable ack!");
+      return;
+    }
+    this->abc_disabled = true;
+    /*
+     * TODO: implement an option to recalibrate sensor, to allow for occasional
+     * manual recalibration
+     */
+  }
```

Comment:
This whole block sounds a bit more like it should go in the `setup()` function. We should probably move it there.

Reply:
I've tried it first but then node hangs for good (had to restart mcu several times before I was able to flash it again) on boot. With UART0  occupied by sensor without hardware to read log from UART1 output, I don't have means to debug it.

I'd still like to move it to setup, but don't know how to fix boot issue.